### PR TITLE
Test compare

### PR DESCRIPTION
**Pull request by jvollhueter**
Current status of the development of a comparison between the results of the test setups and reference files.

### CODE REVIEW

**Comment by jvollhueter** [comment]
In the commit [4143db1](https://github.com/jbathmann/pyMANGA/commit/4143db142d5c21784e657ccec82c48fd15f50242) now there is a test setup where the comparison test is realized by reading the initial conditions for fouling and recruitings (AllSimple_LoadDist.xml) and one (FIXEDSAL_BETTINA.xml) with a manual definition of the numerical values of a variable whose value should actually be determined randomly. So the PR [Seed for random numbers](https://github.com/jbathmann/pyMANGA/pull/57) is already integrated in the commit.

**Comment by jvollhueter** [comment]
Update to the latest work status: Implementation of Test 2 (comparative test) in large tests and adaptation of the example setups (reduction of output files)
still to do:
Sensible handling of output files produced by OGS during a test run

**Comment by jvollhueter** [comment]
Current state of the comparative test, adapted more or less on current version of test setups. I think we should decide now which setups should be tested. Maybe then this stay in folder test and the other could be moved to benchmarks.

**Comment by jvollhueter** [comment]
Current status of the test setups, some are not working yet due to the problems with read from file.

Question: Are the type and scope of the setups okay? Then I would fix the problem with read from file and finish everything.

File: test/smallTests/testSetupsWithoutOGS/smallSetup_1.xml [hunk 1]
```diff
@@ -35,18 +35,18 @@
                 <domain>
                     <x_1> 0 </x_1>
                     <y_1> 0 </y_1>
-                    <x_2> 100 </x_2>
-                    <y_2> 5 </y_2>
+                    <x_2> 20 </x_2>
+                    <y_2> 20 </y_2>
                 </domain>
-                <n_individuals> 200 </n_individuals>
+                <n_individuals> 5 </n_individuals>
```
Comment (jbathmann) [inline test/smallTests/testSetupsWithoutOGS/smallSetup_1.xml:41]: can you plant individuals in each timestep?

File: test/smallTests/testSetupsWithoutOGS/smallSetup_1.xml [hunk 2]
```diff
@@ -35,18 +35,18 @@
...
             </distribution>
         </group>
     </initial_population>
     <tree_time_loop>
         <type> Simple </type>
         <t_start> 0 </t_start>
-        <t_end> 1e9 </t_end>
-        <delta_t> 1e7 </delta_t>
+        <t_end> 86400 </t_end>
```
Comment (jbathmann) [inline test/smallTests/testSetupsWithoutOGS/smallSetup_1.xml:48]: please leave a comment on the timeseps?

File: test/smallTests/testSetupsWithoutOGS/smallSetup_2.xml [hunk 3]
```diff
@@ -1,62 +1,65 @@
...
+                </domain>
+		<x_resolution> 100 </x_resolution>
+		<y_resolution> 100 </y_resolution>
         </aboveground_competition>
         <belowground_competition>
             <type> FixedSalinity </type>
-            <salinity> 0.005 0.095 </salinity>
-            <max_x> 20 </max_x>
-            <min_x> 0 </min_x>
+		<salinity> 0.02 0.02 </salinity>
```
Comment (jbathmann) [inline test/smallTests/testSetupsWithoutOGS/smallSetup_2.xml:17]: please use a gradiet here. I think this is nice to test the bettina behavior for differing salinities

File: test/smallTests/testSetupsWithoutOGS/smallSetup_2.xml [hunk 4]
```diff
@@ -1,62 +1,65 @@
...
-                    <x_1> 0 </x_1>
-                    <y_1> 0 </y_1>
-                    <x_2> 20 </x_2>
-                    <y_2> 20 </y_2>
-                </domain>
-                <n_individuals> 0 </n_individuals>
-                <n_recruitment_per_step> 0 </n_recruitment_per_step>
+                <type> GroupFromFile </type>
+		<n_recruitment_per_step> 5 </n_recruitment_per_step>
+		<filename> test/smallTests/inputFiles/Population_t_0000043200.0.csv </filename>
```
Comment (jbathmann) [inline test/smallTests/testSetupsWithoutOGS/smallSetup_2.xml:34]: does this work?

File: test/smallTests/testSetupsWithoutOGS/smallSetup_3.xml [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        </tree_growth_and_death>
+    </tree_dynamics>
+    <initial_population>
+        <group>
+            <name> Recruiting </name>
+            <species> Avicennia </species>
+            <distribution>
+                <type> GroupFromFile </type>
+		<n_recruitment_per_step> 5 </n_recruitment_per_step>
+		<filename> test/smallTests/inputFiles/Population_t_0000043200.0.csv </filename>
```
Comment (jbathmann) [inline test/smallTests/testSetupsWithoutOGS/smallSetup_3.xml:33]: you could also test the seeding?

File: test/smallTests/testSetupsWithoutOGS/smallSetup_3.xml [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+		<aa> 1 </aa>
+		<bb> 1 </bb>
+		<fmin> 1 </fmin>
+		<salinity> 0.2 </salinity>
+		<domain>
+			<x_1> 0 </x_1>
+                	<y_1> 0 </y_1>
+                	<x_2> 20 </x_2>
+                	<y_2> 20 </y_2>
+			<x_resolution> 100 </x_resolution>
```
Comment (jbathmann) [inline test/smallTests/testSetupsWithoutOGS/smallSetup_3.xml:18]: do you need such a high resolution?

File: test/smallTests/testSetupsWithoutOGS/smallSetup_5.xml [hunk 7]
```diff
@@ -0,0 +1,74 @@
+<?xml version="1.0" encoding="ISO-8859-1"?>
+<MangaProject>
+    <tree_dynamics>
+        <aboveground_competition>
+            <type> SimpleTest </type>
+        </aboveground_competition>
+        <belowground_competition>
+            <type> NetworkHydro </type>
```
Comment (jbathmann) [inline test/smallTests/testSetupsWithoutOGS/smallSetup_5.xml:8]: Please discuss those network setups with @mcwimm

File: test/smallTests/testSetupsWithoutOGS/smallSetup_6.xml [hunk 8]
```diff
@@ -0,0 +1,72 @@
+<?xml version="1.0" encoding="ISO-8859-1"?>
+<MangaProject>
+    <tree_dynamics>
+        <aboveground_competition>
+            <type> SimpleTest </type>
+        </aboveground_competition>
+        <belowground_competition>
+            <type> SimpleHydro </type>
```
Comment (jbathmann) [inline test/smallTests/testSetupsWithoutOGS/smallSetup_6.xml:8]: Please discuss this Setup with either @mcwimm or @zaphod-bx

File: test/smallTests/testSetupsWithoutOGS/smallSetup_7.xml [hunk 9]
```diff
@@ -1,53 +1,55 @@
 <?xml version="1.0" encoding="ISO-8859-1"?>
 <MangaProject>
-    <random_seed> 10 </random_seed>
     <tree_dynamics>
         <aboveground_competition>
             <type> SimpleTest </type>
         </aboveground_competition>
         <belowground_competition>
             <type> SimpleNetwork </type>
```
Comment (jbathmann) [inline test/smallTests/testSetupsWithoutOGS/smallSetup_7.xml:8]: @mcwimm

File: test/smallTests/testWithoutOGS.py [hunk 10]
```diff
@@ -0,0 +1,217 @@
+import sys
```
Comment (jbathmann) [inline test/smallTests/testWithoutOGS.py:1]: what does this scrtipt do?
Please add some explanations in the header of the file

**Comment by jbathmann** [comment]
@mcwimm and @zaphod-bx:
Here, we would like to create some basic test setups, which are tested for every change in the code. To this end, we want to test as many of the code in as few setup files as possible. Could you please have a look at the respective files, if you can suggest some parameters for the setup files?

File: test/smallTests/testSetupsWithoutOGS/smallSetup_1.xml (hunk 2 above)
Comment (jvollhueter) [reply test/smallTests/testSetupsWithoutOGS/smallSetup_1.xml:48]: Done

File: test/smallTests/testSetupsWithoutOGS/smallSetup_2.xml (hunk 3 above)
Comment (jvollhueter) [reply test/smallTests/testSetupsWithoutOGS/smallSetup_2.xml:17]: Done

File: test/smallTests/testSetupsWithoutOGS/smallSetup_1.xml (hunk 1 above)
Comment (jvollhueter) [reply test/smallTests/testSetupsWithoutOGS/smallSetup_1.xml:41]: Done

File: test/smallTests/testSetupsWithoutOGS/smallSetup_3.xml (hunk 6 above)
Comment (jvollhueter) [reply test/smallTests/testSetupsWithoutOGS/smallSetup_3.xml:18]: reduced to 80, this is the smallest possible value

**Comment by jvollhueter** [comment]
the comparison is made by reading the data of the reference file and the result file as pd-dataframe and subtracting them from each other. Then it is checked whether the resulting DF contains only zero values.

The result DF could now be rounded to e.g. 8 digits to prevent false negative test results due to rounding problems.

Next step: Adjustment to PR OGS compitability

File: test/smallTests/testSetupsWithoutOGS/smallSetup_2.xml [hunk 11]
```diff
@@ -1,62 +1,65 @@
...
+                </domain>
+		<x_resolution> 100 </x_resolution>
+		<y_resolution> 100 </y_resolution>
         </aboveground_competition>
         <belowground_competition>
             <type> FixedSalinity </type>
-            <salinity> 0.005 0.095 </salinity>
-            <max_x> 20 </max_x>
-            <min_x> 0 </min_x>
+		<salinity> 0.02 0.03 </salinity>
```
Comment (jbathmann) [inline test/smallTests/testSetupsWithoutOGS/smallSetup_2.xml:17]: number of spaces consisten?

File: test/smallTests/testWithoutOGS.py [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+    print("")
+    print("The testing of all example setups is finished.")
+    print("")
+    print("________________________________________________")
+    print("________________________________________________")
+    print("")
+    print("Report")
+    print("________________________________________________")
+    print("________________________________________________")
+    print("")
```
Comment (jbathmann) [inline test/smallTests/testWithoutOGS.py:146]:
```suggestion
    """
    
    The testing of all example setups is finished.
    ________________________________________________
    
    ...
```

Comment (jbathmann) [reply test/smallTests/testWithoutOGS.py:146]: You can use '"""' to add a block quote

File: test/smallTests/testWithoutOGS.py [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+    if not len(example_setups) == 1:
+        print("The following sample setups have been tested:")
+    else:
+        print("The following sample setup have been tested:")
```
Comment (jbathmann) [inline test/smallTests/testWithoutOGS.py:150]:
```suggestion
The following sample terup(s) haven been tested.
```

Comment (jbathmann) [reply test/smallTests/testWithoutOGS.py:150]: ort ... setup[s] ...

**Comment by jbathmann** [comment]
Cool.
What is still missing?

File: test/smallTests/testSetupsWithoutOGS/smallSetup_2.xml (hunk 11 above)
Comment (jvollhueter) [reply test/smallTests/testSetupsWithoutOGS/smallSetup_2.xml:17]: Done

File: test/smallTests/testWithoutOGS.py (hunk 12 above)
Comment (jvollhueter) [reply test/smallTests/testWithoutOGS.py:146]: Done

File: test/smallTests/testWithoutOGS.py (hunk 13 above)
Comment (jvollhueter) [reply test/smallTests/testWithoutOGS.py:150]: Done

**Comment by jvollhueter** [comment]
Now everything should work theoretically. Unfortunately, reading spatial placements of trees from files did not work. I have now added the manipulation of the random generator to the setups. Do you know why the readout does not work?

Should I adjust the setups somehow? With setup 3 I had the problem that for the second time step for some trees nan values for the stem geometry were output. Trial and error showed that the problem can be solved by changing the parameters aa and bb.

**Comment by mcwimm** [comment]
> Now everything should work theoretically. Unfortunately, reading spatial placements of trees from files did not work. I have now added the manipulation of the random generator to the setups. Do you know why the readout does not work?
> 
> Should I adjust the setups somehow? With setup 3 I had the problem that for the second time step for some trees nan values for the stem geometry were output. Trial and error showed that the problem can be solved by changing the parameters aa and bb.

Do you mean reading in input data ('GroupFromFile' )? Then maybe it's the separator.

**Comment by jvollhueter** [comment]
> > Now everything should work theoretically. Unfortunately, reading spatial placements of trees from files did not work. I have now added the manipulation of the random generator to the setups. Do you know why the readout does not work?
> 
> Do you mean reading in input data ('GroupFromFile' )? Then maybe it's the separator.

The separator should be a comma, right? I have also tried with tabulator, but then comes directly an error message. With the comma no error message comes, but the trees are not placed according to the coordinates in the specified file.

Regardless, the tests should be ready so far and could theoretically be included in the master.

**Pull request merged**